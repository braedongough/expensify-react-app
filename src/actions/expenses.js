import database from '../firebase/firebase'

//ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense: expense
})

export const startAddExpense = (expenseData = {}) => {
    return async (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData
        const expense = { description, note, amount, createdAt }

        const ref = await database.ref('expenses').push(expense)
        dispatch(addExpense({ id: ref.key, ...expense }))
    }
}

//REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense = (id) => {
    return async (dispatch) => {
        await database.ref(`expenses/${id}`).remove()
        dispatch(removeExpense({ id }))
    }
}

//creat startRemoveExpense, receives same args as removeExpense
//"Should remove expense from firebase"
//to verify, assert that snapshot.val() === null  
// use startRemoveExpense in EditExpensePage
//update tests for EditExpensePage

//EDIT_EXPENSE
export const editExpense = (id, update) => ({
    type: 'EDIT_EXPENSE',
    id,
    update
})

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
})

export const startSetExpenses = () => {
    return async (dispatch) => {
        const snapshot = await database.ref('expenses').once('value')
        const expenses = []
        snapshot.forEach((childSnapshot) => {
            expenses.push({
                id: childSnapshot.key,
                ...childSnapshot.val()
            })
        })
        dispatch(setExpenses(expenses))
    }
}