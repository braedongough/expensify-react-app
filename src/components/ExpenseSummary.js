import React from 'react'
import { connect } from 'react-redux'
import selectExpenses from '../selectors/expenses'
import totalExpenses from '../selectors/expenses-total'
import numeral from 'numeral'

export const ExpenseSummary = ({ expenses }) => {
    const total = numeral(totalExpenses(expenses) / 100).format('$0,0.00')
    const expenseCount = expenses.length === 1 ? '1 expense' : `${expenses.length} expenses`
    return (
        <div>
            <p>{`Viewing ${expenseCount}, totaling ${total}`}</p>
        </div>
    )
}

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseSummary)