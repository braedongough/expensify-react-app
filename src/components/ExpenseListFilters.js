import React from 'react'
import { connect } from 'react-redux'
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters'
import { DateRangePicker } from 'react-dates'
import uuid from 'uuid'

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }
    onSortChange = (e) => {
        e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount()
    }
    render() {
        return (
            <div>
                <input
                    type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange} />
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    startDateId={uuid()}
                    endDateId={uuid()}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}
//continue mapping state to props next is endDate
const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => { dispatch(setStartDate(startDate)) },
    setEndDate: (endDate) => { dispatch(setEndDate(endDate)) },
    setTextFilter: (text) => { dispatch(setTextFilter(text)) },
    sortByDate: () => { dispatch(sortByDate()) },
    sortByAmount: () => { dispatch(sortByAmount()) }
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)