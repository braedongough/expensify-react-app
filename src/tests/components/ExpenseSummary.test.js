import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseSummary } from '../../components/ExpenseSummary'
import { expenses } from '../fixtures/expenses'

test('should match ExpenseSummary snapshot for single expense', () => {
    const wrapper = shallow(
        <ExpenseSummary
            expenses={[expenses[0]]}
        />)
    expect(wrapper).toMatchSnapshot()
})

test('should match ExpenseSummary snapshot for multiple expenses', () => {
    const wrapper = shallow(
        <ExpenseSummary
            expenses={expenses}
        />)
    expect(wrapper).toMatchSnapshot()
})