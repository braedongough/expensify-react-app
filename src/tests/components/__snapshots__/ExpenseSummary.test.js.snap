// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match ExpenseSummary snapshot for multiple expenses 1`] = `
<div>
  <p>
    Viewing 3 expenses, totaling $19.02
  </p>
</div>
`;

exports[`should match ExpenseSummary snapshot for single expense 1`] = `
<div>
  <p>
    Viewing 1 expense, totaling $12.34
  </p>
</div>
`;
